src/main/python/treccar/trec_car_classes.py

```diff
@@ -0,0 +1,643 @@
+    """
+    A paragraph.
+    """
+    def __init__(self, para_id, bodies):
+        self.para_id = para_id
+        self.bodies = list(bodies)
+
+    @staticmethod
+    def from_cbor(cbor):
+        assert cbor[0] == 0
+        paragraphId = cbor[1].decode('ascii')
+        return Paragraph(paragraphId, map(ParaBody.from_cbor, cbor[2]))
+
+    def get_text(self):
+        """
+        Get all of the contained text.
+
+        :rtype: str
+        """
+        return ''.join([body.get_text() for body in self.bodies])
+
+    def __str__(self, level=None):
+        return ' '.join(str(body) for body in self.bodies)
+
+class ParaBody(object):
+    """
+    An abstract superclass representing a bit of :class:`Paragraph` content.
+    """
+    @staticmethod
+    def from_cbor(cbor):
+        tag = cbor[0]
+        if tag == 0:
+            return ParaText(cbor[1])
+        elif tag == 1:
+            cbor_ = cbor[1]
+            linkSection = None
+            if len(cbor_[2]) == 1:
+                linkSection = cbor_[2][0]
+            linkTargetId = cbor_[3].decode('ascii')
+            return ParaLink(cbor_[1], linkSection, linkTargetId, cbor_[4])
+        else:
+            assert(False)
+
+    def get_text(self):
+        raise NotImplementedError
+
+    def get_text(self):
+        """
+        Get all of the text within a :class:`ParaBody`.
+
+        :rtype: str
+        """
+        raise NotImplementedError
+
+class ParaText(ParaBody):
+    """
+    A bit of plain text from a paragraph.
+
+    .. attribute:: text
+
+       :rtype: str
+
+       The text
+    """
+    def __init__(self, text):
+        self.text = text
+
+    def get_text(self):
+        return self.text
+
+    def __str__(self, level=None):
+        return self.text
+
+class ParaLink(ParaBody):
+    """
+    A link within a paragraph.
+
+    .. attribute:: page
+
+       :rtype: PageName
+
+       The page name of the link target
+
+    .. attribute:: pageid
+
+       :rtype: PageId
+
+       The link target as trec-car identifer
+
+    .. attribute:: link_section
+
+       :rtype: str
+
+       Section anchor of link target (i.e. the part after the ``#`` in the
+       URL), or ``None``.
+
+    .. attribute:: anchor_text
+
+       :rtype: str
+
+       The anchor text of the link
+    """
+    def __init__(self, page, link_section, pageid, anchor_text):
+        self.page = page
+        self.pageid = pageid
+        self.link_section = link_section
+        self.anchor_text = anchor_text
+
+    def get_text(self):
+        return self.anchor_text
+
+    def __str__(self, level=None):
+        return "[%s](%s)" % (self.anchor_text, self.page)
+
+def _iter_with_header(file, parse, expected_file_types):
+    maybe_hdr = cbor.load(file)
+    if isinstance(maybe_hdr, list) and maybe_hdr[0] == 'CAR':
+        # we have a header
+        file_type = maybe_hdr[1][0]
+        assert file_type in expected_file_types
+
+        # read beginning of variable-length list
```

Comment:
Should we stick to a certain format for the comments, e.g: capitalize first letter, no comma at the end?